### ISSUE TITLE
Get/New-AzRoleAssignment produces error when called as SPN

### ISSUE THREAD

**Issue by jformacek** [Azure/azure-powershell#13573]
## Description
When calling Get/New-AzRoleAssignment as SPN, comand fails and produces this error:
`Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown`

The same command called as user with the same permissions succeeds.

## Steps to reproduce
Let's have AAD Management group MG, User account UA and service principal SPN.
Let's assign Owner role to UA and to SPN on the MG.
Then let's call Get-AzRoleAssignment:
```powershell
Get-AzRoleAssignment -Scope /providers/Microsoft.Management/managementGroups/MG
```
Return the following error:
```
Get-AzRoleAssignment : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
At line:1 char:1
+ Get-AzRoleAssignment -Scope /providers/Microsoft.Management/managemen ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzRoleAssignment], CloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand
```
Error details:
```
$error[0].Exception | select *


Request        : Microsoft.Rest.HttpRequestMessageWrapper
Response       : Microsoft.Rest.HttpResponseMessageWrapper
Body           : Microsoft.Rest.Azure.CloudError
RequestId      : 
Message        : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
Data           : {}
InnerException : 
TargetSite     : Void MoveNext()
StackTrace     :    at Microsoft.Azure.Graph.RBAC.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesAsync>d__6.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d__3.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.GetObjectsByObjectIds(IObjectsOperations operations, GetObjectsParameters parameters)
                    at Microsoft.Azure.Commands.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectId(List`1 objectIds)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleAssignments(IEnumerable`1 assignments, IEnumerable`
                 1 roleDefinitions, AuthorizationClient policyClient, ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignments(FilterRoleAssignmentsOptions options, String cu
                 rrentSubscription, UInt64 first, UInt64 skip)
                    at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
HelpLink       : 
Source         : Microsoft.Azure.Graph.RBAC
HResult        : -2146233088
```
## Environment data
PS Version Table:
```
Name                           Value
----                           -----
PSVersion                      5.1.19041.610
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.19041.610
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

## Module versions

Module versions:

```powershell


    Directory: C:\Users\JiriFormacek\AppData\Local\Apps\SharePointPnPPowerShellOnline\Modules


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Binary     3.17.20... SharePointPnPPowerShellOnline       {Add-PnPAlert, Add-PnPApp, Add-PnPApplicationCustomizer, A...


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Manifest   1.12.0     7Zip4Powershell                     {Expand-7Zip, Compress-7Zip, Get-7Zip, Get-7ZipInformation}  
Script     1.2.2      AiLogging                           {Initialize-AiLogger, Add-AiMetadata, Remove-AiMetadata, R...
Script     2.2.1      Az.Accounts                         {Disable-AzDataCollection, Disable-AzContextAutosave, Enab...
Script     1.9.3      Az.Accounts                         {Disable-AzDataCollection, Disable-AzContextAutosave, Enab...
Script     1.8.1      Az.Accounts                         {Disable-AzDataCollection, Disable-AzContextAutosave, Enab...
Script     1.1.1      Az.Advisor                          {Get-AzAdvisorRecommendation, Enable-AzAdvisorRecommendati...
Script     1.2.0      Az.Aks                              {Get-AzAksCluster, New-AzAksCluster, Remove-AzAksCluster, ...
Script     1.1.1      Az.Aks                              {Get-AzAks, New-AzAks, Remove-AzAks, Import-AzAksCredentia...
Script     1.1.4      Az.AnalysisServices                 {Resume-AzAnalysisServicesServer, Suspend-AzAnalysisServic...
Script     1.1.3      Az.AnalysisServices                 {Resume-AzAnalysisServicesServer, Suspend-AzAnalysisServic...
Script     2.1.0      Az.ApiManagement                    {Add-AzApiManagementApiToGateway, Add-AzApiManagementApiTo...
Script     2.0.1      Az.ApiManagement                    {Add-AzApiManagementApiToProduct, Add-AzApiManagementProdu...
Script     1.1.0      Az.ApplicationInsights              {Get-AzApplicationInsights, New-AzApplicationInsights, Rem...
Script     1.4.0      Az.Automation                       {Get-AzAutomationHybridWorkerGroup, Remove-AzAutomationHyb...
Script     1.3.6      Az.Automation                       {Get-AzAutomationHybridWorkerGroup, Remove-AzAutomationHyb...
Script     3.1.0      Az.Batch                            {Remove-AzBatchAccount, Get-AzBatchAccount, Get-AzBatchAcc...
Script     3.0.0      Az.Batch                            {Remove-AzBatchAccount, Get-AzBatchAccount, Get-AzBatchAcc...
Script     1.0.3      Az.Billing                          {Get-AzBillingInvoice, Get-AzBillingPeriod, Get-AzEnrollme...
Script     1.4.3      Az.Cdn                              {Get-AzCdnProfile, Get-AzCdnProfileSsoUrl, New-AzCdnProfil...
Script     1.5.1      Az.CognitiveServices                {Get-AzCognitiveServicesAccount, Get-AzCognitiveServicesAc...
Script     1.4.0      Az.CognitiveServices                {Get-AzCognitiveServicesAccount, Get-AzCognitiveServicesAc...
Script     4.3.1      Az.Compute                          {Remove-AzAvailabilitySet, Get-AzAvailabilitySet, New-AzAv...
Script     4.1.0      Az.Compute                          {Remove-AzAvailabilitySet, Get-AzAvailabilitySet, New-AzAv...
Script     0.7.0      Az.Consumption                      {Get-AzConsumptionBudget, Get-AzConsumptionMarketplace, Ge...
Script     1.0.3      Az.ContainerInstance                {New-AzContainerGroup, Get-AzContainerGroup, Remove-AzCont...
Script     1.1.1      Az.ContainerRegistry                {New-AzContainerRegistry, Get-AzContainerRegistry, Update-...
Script     1.1.0      Az.DataBoxEdge                      {Get-AzDataBoxEdgeJob, Get-AzDataBoxEdgeDevice, Invoke-AzD...
Script     1.10.0     Az.DataFactory                      {Set-AzDataFactoryV2, Update-AzDataFactoryV2, Get-AzDataFa...
Script     1.8.1      Az.DataFactory                      {Set-AzDataFactoryV2, Update-AzDataFactoryV2, Get-AzDataFa...
Script     1.0.2      Az.DataLakeAnalytics                {Get-AzDataLakeAnalyticsDataSource, New-AzDataLakeAnalytic...
Script     1.2.8      Az.DataLakeStore                    {Get-AzDataLakeStoreTrustedIdProvider, Remove-AzDataLakeSt...
Script     1.0.0      Az.DataShare                        {New-AzDataShareAccount, Get-AzDataShareAccount, Remove-Az...
Script     1.1.0      Az.DeploymentManager                {Get-AzDeploymentManagerArtifactSource, New-AzDeploymentMa...
Script     1.0.0      Az.DesktopVirtualization            {Disconnect-AzWvdUserSession, Get-AzWvdApplication, Get-Az...
Script     1.0.2      Az.DevTestLabs                      {Get-AzDtlAllowedVMSizesPolicy, Get-AzDtlAutoShutdownPolic...
Script     1.1.2      Az.Dns                              {Get-AzDnsRecordSet, New-AzDnsRecordConfig, Remove-AzDnsRe...
Script     1.3.0      Az.EventGrid                        {New-AzEventGridTopic, Get-AzEventGridTopic, Set-AzEventGr...
Script     1.2.3      Az.EventGrid                        {New-AzEventGridTopic, Get-AzEventGridTopic, Set-AzEventGr...
Script     1.5.0      Az.EventHub                         {New-AzEventHubNamespace, Get-AzEventHubNamespace, Set-AzE...
Script     1.4.3      Az.EventHub                         {New-AzEventHubNamespace, Get-AzEventHubNamespace, Set-AzE...
Script     1.6.1      Az.FrontDoor                        {New-AzFrontDoor, Get-AzFrontDoor, Set-AzFrontDoor, Remove...
Script     1.5.0      Az.FrontDoor                        {New-AzFrontDoor, Get-AzFrontDoor, Set-AzFrontDoor, Remove...
Script     1.0.1      Az.Functions                        {Get-AzFunctionApp, Get-AzFunctionAppAvailableLocation, Ge...
Script     1.0.0      Az.Functions                        {Get-AzFunctionApp, Get-AzFunctionAppAvailableLocation, Ge...
Script     3.5.0      Az.HDInsight                        {Get-AzHDInsightJob, New-AzHDInsightSqoopJobDefinition, Wa...
Script     3.2.0      Az.HDInsight                        {Get-AzHDInsightJob, New-AzHDInsightSqoopJobDefinition, Wa...
Script     1.1.0      Az.HealthcareApis                   {New-AzHealthcareApisService, Remove-AzHealthcareApisServi...
Script     1.0.2      Az.HealthcareApis                   {New-AzHealthcareApisService, Remove-AzHealthcareApisServi...
Script     2.5.0      Az.IotHub                           {Add-AzIotHubKey, Get-AzIotHubEventHubConsumerGroup, Get-A...
Script     2.1.0      Az.KeyVault                         {Add-AzKeyVaultCertificate, Update-AzKeyVaultCertificate, ...
Script     2.0.0      Az.KeyVault                         {Add-AzKeyVaultCertificate, Update-AzKeyVaultCertificate, ...
Script     1.3.2      Az.LogicApp                         {Get-AzIntegrationAccountAgreement, Get-AzIntegrationAccou...
Script     1.1.3      Az.MachineLearning                  {Move-AzMlCommitmentAssociation, Get-AzMlCommitmentAssocia...
Script     1.1.0      Az.Maintenance                      {Get-AzApplyUpdate, Get-AzConfigurationAssignment, Get-AzM...
Script     1.0.0      Az.Maintenance                      {Get-AzApplyUpdate, Get-AzConfigurationAssignment, Get-AzM...
Script     1.1.0      Az.ManagedServices                  {Get-AzManagedServicesAssignment, New-AzManagedServicesAss...
Script     1.0.2      Az.ManagedServices                  {Get-AzManagedServicesAssignment, New-AzManagedServicesAss...
Script     1.0.2      Az.MarketplaceOrdering              {Get-AzMarketplaceTerms, Set-AzMarketplaceTerms}             
Script     1.1.1      Az.Media                            {Sync-AzMediaServiceStorageKey, Set-AzMediaServiceKey, Get...
Script     2.1.0      Az.Monitor                          {Get-AzMetricDefinition, Get-AzMetric, Remove-AzLogProfile...
Script     2.0.0      Az.Monitor                          {Get-AzMetricDefinition, Get-AzMetric, Remove-AzLogProfile...
Script     3.3.0      Az.Network                          {Add-AzApplicationGatewayAuthenticationCertificate, Get-Az...
Script     3.0.0      Az.Network                          {Add-AzApplicationGatewayAuthenticationCertificate, Get-Az...
Script     1.1.1      Az.NotificationHubs                 {Get-AzNotificationHub, Get-AzNotificationHubAuthorization...
Script     2.3.0      Az.OperationalInsights              {New-AzOperationalInsightsAzureActivityLogDataSource, New-...
Script     2.1.0      Az.OperationalInsights              {New-AzOperationalInsightsAzureActivityLogDataSource, New-...
Script     1.3.1      Az.PolicyInsights                   {Get-AzPolicyEvent, Get-AzPolicyState, Get-AzPolicyStateSu...
Script     1.1.2      Az.PowerBIEmbedded                  {Remove-AzPowerBIWorkspaceCollection, Get-AzPowerBIWorkspa...
Script     1.0.3      Az.PrivateDns                       {Get-AzPrivateDnsZone, Remove-AzPrivateDnsZone, Set-AzPriv...
Script     2.11.1     Az.RecoveryServices                 {Get-AzRecoveryServicesBackupProperty, Get-AzRecoveryServi...
Script     2.9.1      Az.RecoveryServices                 {Get-AzRecoveryServicesBackupProperty, Get-AzRecoveryServi...
Script     1.2.1      Az.RedisCache                       {Remove-AzRedisCachePatchSchedule, New-AzRedisCacheSchedul...
Script     1.0.3      Az.Relay                            {New-AzRelayNamespace, Get-AzRelayNamespace, Set-AzRelayNa...
Script     3.0.1      Az.Resources                        {Get-AzProviderOperation, Remove-AzRoleAssignment, Get-AzR...
Script     2.5.0      Az.Resources                        {Get-AzProviderOperation, Remove-AzRoleAssignment, Get-AzR...
Script     2.1.0      Az.Resources                        {Get-AzProviderOperation, Remove-AzRoleAssignment, Get-AzR...
Script     1.4.1      Az.ServiceBus                       {New-AzServiceBusNamespace, Get-AzServiceBusNamespace, Set...
Script     2.1.0      Az.ServiceFabric                    {Add-AzServiceFabricClientCertificate, Add-AzServiceFabric...
Script     1.2.0      Az.SignalR                          {New-AzSignalR, Get-AzSignalR, Get-AzSignalRKey, New-AzSig...
Script     1.1.1      Az.SignalR                          {New-AzSignalR, Get-AzSignalR, Get-AzSignalRKey, New-AzSig...
Script     2.9.1      Az.Sql                              {Get-AzSqlDatabaseTransparentDataEncryption, Get-AzSqlData...
Script     2.7.0      Az.Sql                              {Get-AzSqlDatabaseTransparentDataEncryption, Get-AzSqlData...
Script     1.1.0      Az.SqlVirtualMachine                {New-AzSqlVM, Get-AzSqlVM, Update-AzSqlVM, Remove-AzSqlVM...}
Script     2.5.0      Az.Storage                          {Get-AzStorageAccount, Get-AzStorageAccountKey, New-AzStor...
Script     2.1.0      Az.Storage                          {Get-AzStorageAccount, Get-AzStorageAccountKey, New-AzStor...
Script     1.3.0      Az.StorageSync                      {Invoke-AzStorageSyncCompatibilityCheck, New-AzStorageSync...
Script     1.2.3      Az.StorageSync                      {Invoke-AzStorageSyncCompatibilityCheck, New-AzStorageSync...
Script     1.0.1      Az.StreamAnalytics                  {Get-AzStreamAnalyticsFunction, Get-AzStreamAnalyticsDefau...
Script     1.0.0      Az.Support                          {Get-AzSupportService, Get-AzSupportProblemClassification,...
Script     1.0.4      Az.TrafficManager                   {Add-AzTrafficManagerCustomHeaderToEndpoint, Remove-AzTraf...
Script     1.11.0     Az.Websites                         {Get-AzAppServicePlan, Set-AzAppServicePlan, New-AzAppServ...
Script     1.9.0      Az.Websites                         {Get-AzAppServicePlan, Set-AzAppServicePlan, New-AzAppServ...
Binary     2.0.2.105  AzureADPreview                      {Add-AzureADApplicationOwner, Get-AzureADApplication, Get-...
Script     4.2.0      CosmosDB                            {Get-CosmosDbAccount, Get-CosmosDbAccountConnectionString,...
Script     0.3582.0   ExchangeOnlineManagement            {Get-EXOCasMailbox, Get-EXOMailbox, Get-EXOMailboxFolderPe...
Script     0.3374.10  ExchangeOnlineManagement            {Get-EXOCasMailbox, Get-EXOMailbox, Get-EXOMailboxFolderPe...
Script     1.0.1      Microsoft.PowerShell.Operation.V... {Get-OperationValidation, Invoke-OperationValidation}        
Script     4.10.0.1   MSAL.PS                             {Add-MsalClientApplication, Clear-MsalTokenCache, Enable-M...
Manifest   1.1.183.57 MSOnline                            {Get-MsolDevice, Remove-MsolDevice, Enable-MsolDevice, Dis...
Script     1.4.6      PackageManagement                   {Find-Package, Get-Package, Get-PackageProvider, Get-Packa...
Script     1.4.5      PackageManagement                   {Find-Package, Get-Package, Get-PackageProvider, Get-Packa...
Binary     1.0.0.1    PackageManagement                   {Find-Package, Get-Package, Get-PackageProvider, Get-Packa...
Script     3.4.0      Pester                              {Describe, Context, It, Should...}                           
Script     2.2.3      PowerShellGet                       {Find-Command, Find-DSCResource, Find-Module, Find-RoleCap...
Script     2.2.2      PowerShellGet                       {Find-Command, Find-DSCResource, Find-Module, Find-RoleCap...
Script     1.0.0.1    PowerShellGet                       {Install-Module, Find-Module, Save-Module, Update-Module...} 
Script     2.0.0      PSReadline                          {Get-PSReadLineKeyHandler, Set-PSReadLineKeyHandler, Remov...
Script     2.1.0      S.DS.P                              {Find-LdapObject, Get-RootDSE, Get-LdapConnection, Edit-Ld...
Script     2.0.6      S.DS.P                              {Find-LDAPObject, Get-RootDSE, Get-LdapConnection, Edit-Ld...
Script     2.0.2      S.DS.P                              {Find-LDAPObject, Get-RootDSE, Get-LdapConnection, Edit-Ld...
Binary     3.17.20... SharePointPnPPowerShellOnline       {Add-PnPAlert, Add-PnPApp, Add-PnPApplicationCustomizer, A...
Script     1.0.3      WHfBTools                           {Get-ADWHfBKeys, Remove-ADWHfBKey, Get-AzureADWHfBKeys, Re...


    Directory: C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Binary     7.7.4.1    AdmPwd.PS                           {Get-AdmPwdEnvironmentStatus, Get-AdmPwdPdsLicenseParamete...
Manifest   1.0.0.0    AppBackgroundTask                   {Disable-AppBackgroundTaskDiagnosticLog, Enable-AppBackgro...
Manifest   2.0.0.0    AppLocker                           {Get-AppLockerFileInformation, Get-AppLockerPolicy, New-Ap...
Manifest   1.0.0.0    AppvClient                          {Add-AppvClientConnectionGroup, Add-AppvClientPackage, Add...
Manifest   2.0.1.0    Appx                                {Add-AppxPackage, Get-AppxPackage, Get-AppxPackageManifest...
Script     1.0.0.0    AssignedAccess                      {Clear-AssignedAccess, Get-AssignedAccess, Set-AssignedAcc...
Manifest   1.0.0.0    BitLocker                           {Unlock-BitLocker, Suspend-BitLocker, Resume-BitLocker, Re...
Manifest   2.0.0.0    BitsTransfer                        {Add-BitsFile, Complete-BitsTransfer, Get-BitsTransfer, Re...
Manifest   1.0.0.0    BranchCache                         {Add-BCDataCacheExtension, Clear-BCCache, Disable-BC, Disa...
Manifest   1.0.0.0    CimCmdlets                          {Get-CimAssociatedInstance, Get-CimClass, Get-CimInstance,...
Manifest   1.0        ConfigCI                            {Get-SystemDriver, New-CIPolicyRule, New-CIPolicy, Get-CIP...
Manifest   1.0        ConfigDefender                      {Get-MpPreference, Set-MpPreference, Add-MpPreference, Rem...
Manifest   1.0        Defender                            {Get-MpPreference, Set-MpPreference, Add-MpPreference, Rem...
Manifest   1.0.2.0    DeliveryOptimization                {Delete-DeliveryOptimizationCache, Set-DeliveryOptimizatio...
Manifest   1.0.0.0    DirectAccessClientComponents        {Disable-DAManualEntryPointSelection, Enable-DAManualEntry...
Script     3.0        Dism                                {Add-AppxProvisionedPackage, Add-WindowsDriver, Add-Window...
Manifest   1.0.0.0    DnsClient                           {Resolve-DnsName, Clear-DnsClientCache, Get-DnsClient, Get...
Manifest   1.0.0.0    EventTracingManagement              {Start-EtwTraceSession, New-EtwTraceSession, Get-EtwTraceS...
Manifest   1.0.0.0    HgsClient                           {Get-HgsAttestationBaselinePolicy, Get-HgsClientConfigurat...
Manifest   1.0.0.0    HgsDiagnostics                      {New-HgsTraceTarget, Get-HgsTrace, Get-HgsTraceFileData, T...
Manifest   1.0.0.1    HostNetworkingService               {Remove-HnsNamespace, Remove-HnsEndpoint, Get-HnsEndpoint,...
Manifest   2.0.0.0    International                       {Get-WinDefaultInputMethodOverride, Set-WinDefaultInputMet...
Manifest   1.0.0.0    iSCSI                               {Get-IscsiTargetPortal, New-IscsiTargetPortal, Remove-Iscs...
Script     1.0.0.0    ISE                                 {New-IseSnippet, Import-IseSnippet, Get-IseSnippet}          
Manifest   1.0.0.0    Kds                                 {Add-KdsRootKey, Get-KdsRootKey, Test-KdsRootKey, Set-KdsC...
Manifest   1.0.1.0    Microsoft.PowerShell.Archive        {Compress-Archive, Expand-Archive}                           
Manifest   3.0.0.0    Microsoft.PowerShell.Diagnostics    {Get-WinEvent, Get-Counter, Import-Counter, Export-Counter...
Manifest   3.0.0.0    Microsoft.PowerShell.Host           {Start-Transcript, Stop-Transcript}                          
Manifest   1.0.0.0    Microsoft.PowerShell.LocalAccounts  {Add-LocalGroupMember, Disable-LocalUser, Enable-LocalUser...
Manifest   3.1.0.0    Microsoft.PowerShell.Management     {Add-Content, Clear-Content, Clear-ItemProperty, Join-Path...
Script     1.0        Microsoft.PowerShell.ODataUtils     Export-ODataEndpointProxy                                    
Manifest   3.0.0.0    Microsoft.PowerShell.Security       {Get-Acl, Set-Acl, Get-PfxCertificate, Get-Credential...}    
Manifest   3.1.0.0    Microsoft.PowerShell.Utility        {Format-List, Format-Custom, Format-Table, Format-Wide...}   
Manifest   3.0.0.0    Microsoft.WSMan.Management          {Disable-WSManCredSSP, Enable-WSManCredSSP, Get-WSManCredS...
Manifest   1.0        MMAgent                             {Disable-MMAgent, Enable-MMAgent, Set-MMAgent, Get-MMAgent...
Manifest   1.0.0.0    MsDtc                               {New-DtcDiagnosticTransaction, Complete-DtcDiagnosticTrans...
Manifest   2.0.0.0    NetAdapter                          {Disable-NetAdapter, Disable-NetAdapterBinding, Disable-Ne...
Manifest   1.0.0.0    NetConnection                       {Get-NetConnectionProfile, Set-NetConnectionProfile}         
Manifest   1.0.0.0    NetEventPacketCapture               {New-NetEventSession, Remove-NetEventSession, Get-NetEvent...
Manifest   2.0.0.0    NetLbfo                             {Add-NetLbfoTeamMember, Add-NetLbfoTeamNic, Get-NetLbfoTea...
Manifest   1.0.0.0    NetNat                              {Get-NetNat, Get-NetNatExternalAddress, Get-NetNatStaticMa...
Manifest   2.0.0.0    NetQos                              {Get-NetQosPolicy, Set-NetQosPolicy, Remove-NetQosPolicy, ...
Manifest   2.0.0.0    NetSecurity                         {Get-DAPolicyChange, New-NetIPsecAuthProposal, New-NetIPse...
Manifest   1.0.0.0    NetSwitchTeam                       {New-NetSwitchTeam, Remove-NetSwitchTeam, Get-NetSwitchTea...
Manifest   1.0.0.0    NetTCPIP                            {Get-NetIPAddress, Get-NetIPInterface, Get-NetIPv4Protocol...
Manifest   1.0.0.0    NetworkConnectivityStatus           {Get-DAConnectionStatus, Get-NCSIPolicyConfiguration, Rese...
Manifest   1.0.0.0    NetworkSwitchManager                {Disable-NetworkSwitchEthernetPort, Enable-NetworkSwitchEt...
Manifest   1.0.0.0    NetworkTransition                   {Add-NetIPHttpsCertBinding, Disable-NetDnsTransitionConfig...
Manifest   1.0.0.0    PcsvDevice                          {Get-PcsvDevice, Start-PcsvDevice, Stop-PcsvDevice, Restar...
Binary     1.0.0.0    PersistentMemory                    {Get-PmemDisk, Get-PmemPhysicalDevice, Get-PmemUnusedRegio...
Manifest   1.0.0.0    PKI                                 {Add-CertificateEnrollmentPolicyServer, Export-Certificate...
Manifest   1.0.0.0    PnpDevice                           {Get-PnpDevice, Get-PnpDeviceProperty, Enable-PnpDevice, D...
Manifest   1.1        PrintManagement                     {Add-Printer, Add-PrinterDriver, Add-PrinterPort, Get-Prin...
Binary     1.0.11     ProcessMitigations                  {Get-ProcessMitigation, Set-ProcessMitigation, ConvertTo-P...
Script     3.0        Provisioning                        {Install-ProvisioningPackage, Export-ProvisioningPackage, ...
Manifest   1.1        PSDesiredStateConfiguration         {Set-DscLocalConfigurationManager, Start-DscConfiguration,...
Script     1.0.0.0    PSDiagnostics                       {Disable-PSTrace, Disable-PSWSManCombinedTrace, Disable-WS...
Binary     1.1.0.0    PSScheduledJob                      {New-JobTrigger, Add-JobTrigger, Remove-JobTrigger, Get-Jo...
Manifest   2.0.0.0    PSWorkflow                          {New-PSWorkflowExecutionOption, New-PSWorkflowSession, nwsn} 
Manifest   1.0.0.0    PSWorkflowUtility                   Invoke-AsWorkflow                                            
Manifest   2.0.0.0    SecureBoot                          {Confirm-SecureBootUEFI, Set-SecureBootUEFI, Get-SecureBoo...
Manifest   1.0.0.0    ScheduledTasks                      {Get-ScheduledTask, Set-ScheduledTask, Register-ScheduledT...
Manifest   2.0.0.0    SmbShare                            {Get-SmbShare, Remove-SmbShare, Set-SmbShare, Block-SmbSha...
Manifest   2.0.0.0    SmbWitness                          {Get-SmbWitnessClient, Move-SmbWitnessClient, gsmbw, msmbw...
Manifest   1.0.0.0    StartLayout                         {Export-StartLayout, Import-StartLayout, Export-StartLayou...
Manifest   2.0.0.0    Storage                             {Add-InitiatorIdToMaskingSet, Add-PartitionAccessPath, Add...
Manifest   1.0.0.0    StorageBusCache                     {Clear-StorageBusDisk, Disable-StorageBusCache, Disable-St...
Manifest   2.0.0.0    TLS                                 {New-TlsSessionTicketKey, Enable-TlsSessionTicketKey, Disa...
Manifest   1.0.0.0    TroubleshootingPack                 {Get-TroubleshootingPack, Invoke-TroubleshootingPack}        
Manifest   2.0.0.0    TrustedPlatformModule               {Get-Tpm, Initialize-Tpm, Clear-Tpm, Unblock-Tpm...}         
Binary     2.1.639.0  UEV                                 {Clear-UevConfiguration, Clear-UevAppxPackage, Restore-Uev...
Manifest   2.0.0.0    VpnClient                           {Add-VpnConnection, Set-VpnConnection, Remove-VpnConnectio...
Manifest   1.0.0.0    Wdac                                {Get-OdbcDriver, Set-OdbcDriver, Get-OdbcDsn, Add-OdbcDsn...}
Manifest   2.0.0.0    Whea                                {Get-WheaMemoryPolicy, Set-WheaMemoryPolicy}                 
Manifest   1.0.0.0    WindowsDeveloperLicense             {Get-WindowsDeveloperLicense, Unregister-WindowsDeveloperL...
Script     1.0        WindowsErrorReporting               {Enable-WindowsErrorReporting, Disable-WindowsErrorReporti...
Manifest   1.0.0.0    WindowsSearch                       {Get-WindowsSearchSetting, Set-WindowsSearchSetting}         
Manifest   1.0.0.0    WindowsUpdate                       Get-WindowsUpdateLog                                         
Manifest   1.0.0.2    WindowsUpdateProvider               {Get-WUAVersion, Get-WULastInstallationDate, Get-WULastSca...


    Directory: C:\Program Files\Common Files\Skype for Business Online\Modules


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Manifest   7.0.0.0    LyncOnlineConnector                                                                              
Script     7.0.0.0    SkypeOnlineConnector                {Get-CsOnlinePowerShellEndpoint, Get-CsOnlinePowerShellAcc...


    Directory: C:\Program Files\SharePoint Online Management Shell


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Binary     16.0.20... Microsoft.Online.SharePoint.Powe... {Register-SPODataEncryptionPolicy, Update-SPODataEncryptio...


    Directory: C:\Program Files (x86)\Microsoft Azure Information Protection\Powershell


ModuleType Version    Name                                ExportedCommands                                             
---------- -------    ----                                ----------------                                             
Binary     2.8.85.0   AzureInformationProtection          {Get-MIPNetworkDiscoveryStatus, Get-MIPNetworkDiscoveryCon...


```

## Debug output

Too long to paste here (~10K lines), stored in file behind link below:

[Debug output](https://gcstoragedownload.blob.core.windows.net/download/get-AzRoleAssignment_DEBUG.txt)

## Error output

Resolve-AzError output:
```
DEBUG: 8:54:46 - ResolveError begin processing with ParameterSet 'AnyErrorParameterSet'.
DEBUG: 8:54:46 - using account id 'ef96b166-d851-45dc-87bb-03e7aa9a28e9'...
WARNING: Upcoming breaking changes in the cmdlet 'Resolve-AzError' :
The `Resolve-Error` alias will be removed in a future release.  Please change any scripts that use this alias to use `Resolve-AzError` instead.
Note : Go to https://aka.ms/azps-changewarnings for steps to suppress this breaking change warning, and other information on breaking changes in Azure PowerShell.


   HistoryId: 20


RequestId      :
Message        : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
ServerMessage  : :  (System.Collections.Generic.List`1[Microsoft.Rest.Azure.CloudError])
ServerResponse : {Forbidden}
RequestMessage : {POST https://graph.windows.net/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/getObjectsByObjectIds?api-version=1.6}
InvocationInfo : {Get-AzRoleAssignment}
Line           : Get-AzRoleAssignment -Scope /providers/Microsoft.Management/managementGroups/MG
Position       : At line:1 char:1
                 + Get-AzRoleAssignment -Scope /providers/Microsoft.Management/managemen ...
                 + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
StackTrace     :    at Microsoft.Azure.Graph.RBAC.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesAsync>d__6.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d__3.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.GetObjectsByObjectIds(IObjectsOperations operations, GetObjectsParameters parameters)
                    at Microsoft.Azure.Commands.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectId(List`1 objectIds)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleAssignments(IEnumerable`1 assignments, IEnumerable`1 roleDefinitions, Author
                 izationClient policyClient, ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignments(FilterRoleAssignmentsOptions options, String currentSubscription, UInt64
                  first, UInt64 skip)
                    at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
HistoryId      : 20


The Azure PowerShell team is listening, please let us know how we are doing: https://aka.ms/azpssurvey?Q_CHL=ERROR.

DEBUG: AzureQoSEvent: CommandName - Resolve-AzError; IsSuccess - True; Duration - 00:00:00.0978340;
DEBUG: Finish sending metric.
DEBUG: 8:54:47 - ResolveError end processing.

```

**Comment by stefanpetter** [comment]
I have the exact same problem! I am trying to assign a policy with an identity and then give that identity the permissions it needs. Logged in as a service principaI that is owner on the subscription. 

```
$roleDefinitionIds = $policyDefinition.Properties.policyRule.then.details.roleDefinitionIds

$assignment = New-AzPolicyAssignment `
-PolicyDefinition $policyDefinition `
-DisplayName $policy.displayName `
-Scope ("/subscriptions/"+$subscription.Id) `
-Name $policy.assignmentName `
-PolicyParameterObject (Invoke-Expression $policy.parameters) `
-Location $resourceLocation `
-AssignIdentity

#If we don't add this sleep, the role assignment will fail, because the identity will not exist
Start-Sleep -s 15

if ($roleDefinitionIds.Count -gt 0)
{
    $roleDefinitionIds | ForEach-Object {
        $roleDefId = $_.Split("/") | Select-Object -Last 1
        New-AzRoleAssignment -ObjectId $assignment.Identity.PrincipalId -Scope ("/subscriptions/"+$subscription.Id) -RoleDefinitionId $roleDefId
    }
}
```
The role assignment results in:
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/57794508/100076001-fb125780-2e40-11eb-976c-36f2e2f1460a.png)

I ran a HTTP listener on my device when I tried to assign the role. The error states the following: 
![MicrosoftTeams-image](https://user-images.githubusercontent.com/57794508/100075357-23e61d00-2e40-11eb-9173-b4dfc14f89e3.png)

I hope this helps with debugging!

**Comment by jformacek** [comment]
And meanwhile, we found a root cause and fix/workaround: seems that command tries to talk to legacy AAD Graph API and resolve objectIDs, and this return Forbidden error
We added permission `Directory.Read.All` application permission to app registration and it started working.
It's kind of shame that Az cmdlets still use the legacy API that has been discouraged for few years now. Also, the documentation could be more specific in terms of level of permission needed to be able to successfully run the cmdlet.

**Comment by stefanpetter** [comment]
> And meanwhile, we found a root cause and fix/workaround: seems that command tries to talk to legacy AAD Graph API and resolve objectIDs, and this return Forbidden error
> We added permission `Directory.Read.All` application permission to app registration and it started working.
> It's kind of shame that Az cmdlets still use the legacy API that has been discouraged for few years now. Also, the documentation could be more specific in terms of level of permission needed to be able to successfully run the cmdlet.

This worked like a charm. Thank you very much for sharing. Hopefully, they'll fix this asap.

**Comment by dingmeng-xue** [comment]
@jformacek , thanks for your reporting and solution. Could you share us more about legacy AAD graph API? We checked the latest Azure.Graph.RBAC which is using 1.6 as well. 

Are you thinking we should use graph.microsoft.com?

**Comment by jformacek** [comment]
Hello @dingmeng-xue dingmeng-xue,
I can see you're calling Legacy AAD Graph endpoint: `https://graph.windows.net/<tenantId>/getObjectsByObjectIds?api-version=1.6`. Direct equivalent is MS Graph endpoint `https://graph.microsoft.com/v1.0/directoryObjects/getByIds` if you wanted to switch to MS Graph.
Other question is why and how you call the AAD Graph: I can see that Get-AzRoleAssignment return array of [Microsoft.Azure.Commands.Resources.Models.Authorization.PSRoleAssignment](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.commands.resources.models.authorization.psroleassignment)  that contains (among others) fields DisplayName, ObjectType a SignInName that are filled by properties of directoryObject that has RBAC role assigned, so this is what you call AAD Graph for. However, you ask AAD Graph to return all information about directory objects, so for each RBAC entry, you get a huge object (sample below) and use really small piece information of it - not a best approach from scalability/performance point of view.  

So my recommendation would be that you could look to switching to MS Graph from AAD Graph, and use $select filter with query to limit set of attributes returned to only those you need to populated output object of the command
Thanks,
Jiri

---

Sample object (user account in this case) returned for each directory object present in RBAC on given scope - please compare amount of information returned to those you return with `Microsoft.Azure.Commands.Resources.Models.Authorization.PSRoleAssignment`
```json
    {
      "odata.type": "Microsoft.DirectoryServices.User",
      "objectType": "User",
      "objectId": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
      "deletionTimestamp": null,
      "accountEnabled": true,
      "ageGroup": null,
      "assignedLicenses": [
        {
          "disabledPlans": [],
          "skuId": "84a661c4-e949-4bd2-a560-ed7766fcaf2b"
        },
        {
          "disabledPlans": [],
          "skuId": "a403ebcc-fae0-4ca2-8c8c-7a907fd6c235"
        }
      ],
      "assignedPlans": [
        {
          "assignedTimestamp": "2020-09-08T09:19:55Z",
          "capabilityStatus": "Enabled",
          "service": "MultiFactorService",
          "servicePlanId": "8a256a2b-b617-496d-b51b-e76466e88db0"
        },
        {
          "assignedTimestamp": "2020-09-08T09:19:55Z",
          "capabilityStatus": "Enabled",
          "service": "AADPremiumService",
          "servicePlanId": "eec0eb4f-6444-4f95-aba0-50c24d67f998"
        },
        {
          "assignedTimestamp": "2020-09-08T09:19:55Z",
          "capabilityStatus": "Enabled",
          "service": "Adallom",
          "servicePlanId": "932ad362-64a8-4783-9106-97849a1a30b9"
        },
        {
          "assignedTimestamp": "2020-09-08T09:19:55Z",
          "capabilityStatus": "Enabled",
          "service": "AADPremiumService",
          "servicePlanId": "41781fb2-bc02-4b7c-bd55-b576c07bb09d"
        },
        {
          "assignedTimestamp": "2020-04-09T11:13:10Z",
          "capabilityStatus": "Enabled",
          "service": "AzureAnalysis",
          "servicePlanId": "2049e525-b859-401b-b2a0-e0a31c4b1fe4"
        }
      ],
      "city": null,
      "companyName": null,
      "consentProvidedForMinor": null,
      "country": null,
      "createdDateTime": "2019-09-04T12:23:22Z",
      "creationType": null,
      "department": null,
      "dirSyncEnabled": null,
      "displayName": "xxxxxx",
      "employeeId": null,
      "facsimileTelephoneNumber": null,
      "givenName": null,
      "immutableId": "xxxxxxxxxxx",
      "isCompromised": null,
      "jobTitle": null,
      "lastDirSyncTime": null,
      "legalAgeGroupClassification": null,
      "mail": null,
      "mailNickname": "xxxxxxxxx",
      "mobile": null,
      "onPremisesDistinguishedName": null,
      "onPremisesSecurityIdentifier": null,
      "otherMails": [
        "xxx@xxx.xx"
      ],
      "passwordPolicies": "None",
      "passwordProfile": null,
      "physicalDeliveryOfficeName": null,
      "postalCode": null,
      "preferredLanguage": "en-GB",
      "provisionedPlans": [],
      "provisioningErrors": [],
      "proxyAddresses": [],
      "refreshTokensValidFromDateTime": null,
      "showInAddressList": null,
      "signInNames": [],
      "sipProxyAddress": null,
      "state": null,
      "streetAddress": null,
      "surname": null,
      "telephoneNumber": null,
      "thumbnailPhoto@odata.mediaEditLink": "directoryObjects/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/Microsoft.DirectoryServices.User/thumbnailPhoto",
      "usageLocation": "XX",
      "userIdentities": [],
      "userPrincipalName": "xxx@xxx.xx",
      "userState": null,
      "userStateChangedOn": null,
      "userType": "Member"
    }
```

**Comment by jformacek** [comment]
Also, option could be that you do not resolve principal references in RBAC at all, and let the callers resolve it on their own. Could help separate duties and make permission model clearer/simpler.
BTW I would be interested in knowing how often caller calls into Graph to get more information about principal than returned in PSRoleAssignment object? I guess this is something that telemetry could help with... It could reveal that your roundtrip to Graph is useless because callers tend to call the Graph anyway to retrieve more about principals?
Jiri

**Comment by orsharon7** [comment]
Same here.

Get-AzRoleAssignment returns:
`Get-AzRoleAssignment : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
At line:1 char:1
+ Get-AzRoleAssignment -ObjectId $($Identity.PrincipalId)  -RoleDefinit ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzRoleAssignment], CloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand`

While Set-AzRoleAssignment returns the same error but the actions actually works.

**Comment by rickvanc** [comment]
We also experience the same problem. 
We have a SP that is nested with a group and that group is given the rights that normally would be enough. But still we get this error.

**Comment by cutecycle** [comment]
For `New-AzRoleAssignment`: Some folks won't be able to easily get `Graph.Read.All` for their SPN, and may only have `Graph.Read.User` and the ability to perform role assignments in a limited context. For them, the issue is `ToPSRoleAssignments` -> `POST getObjectsByObjectId` (even for a single object) -> `403`. 

https://github.com/Azure/azure-powershell/blob/63193bf2bdb03ae30a175570fce0095864ddb093/src/Resources/Resources/ActiveDirectory/Models/ActiveDirectoryClient.cs#L280

The assignment still succeeds, though: ToPSRoleAssignments is a presentation function that occurs after the role assignment.
At some point down the chain, the 403 status of the graph REST request CloudException is swallowed and you get back the blank CloudException.

Workaround, though not ideal: swallow the cloudexception from New-AzRoleAssignment yourself and find a permissioned, or out of band, way to trust, but verify the result.

**Comment by ivanjrt** [comment]
Does anybody know how to fix the issue with  a 'Managed Identity' ?

**Comment by JohnLudlow** [comment]
I am also seeing this issue - I'm trying to script a role assignment to give a machine admin account access to a blob so I can use `Set-AzVmDscExtension`.

**Comment by Mslatersuncor** [comment]
After fighting this for a number of weeks now I got it to work with having both MS Graph and Legacy AAD Graph API permissions configured. This is the same as what jformacek reported back in November

![image](https://user-images.githubusercontent.com/57813478/123472941-1864fb00-d5b5-11eb-9526-4a068b8c7697.png)

**Comment by gilmoreg** [comment]
Just noticed that the Azure CLI has a workaround for this same issue:

https://docs.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest

--assignee-object-id
Use this parameter instead of '--assignee' to bypass Graph API invocation in case of insufficient privileges. This parameter only works with object ids for users, groups, service principals, and managed identities. For managed identities use the principal id. For service principals, use the object id and not the app id.

**Comment by jikuja** [comment]
How this is still not fixed or properly documented.

The impact of the problem is huge and causes issues with DevOps pipelines, automation, and probably with ARM deployment scripts.

--------

One more workaround. Provide a set of arguments that avoids calling Graph APIs. E.g.:

```powershell
New-AzRoleAssignment `
-ObjectId $id `
-ObjectType "Group" `
-RoleDefinitionId "00482a5a-887f-4fb3-b363-3b7fe8e74483" `
-Scope "/subscriptions/${Env:SUBSCRIPTIONID}/resourcegroups/${Env:RESOURCEGROUPNAME}"
```

To use `-RoleDefinitionId` you must use also `-scope`. Using `-ObjectType` might be the final missing piece. This of course works only for `New-AzRoleAssignment` Cmdlet.

This helps only with scenarios when Ids are known.

**Comment by hannwang1** [comment]
any update from the owner team of New-AzRoleAssignment, Get-AzRoleAssignment Cmdlet about fixing this issue? This is blocking our release pipeline.  Our release SPN cannot get graph directory.Read.All.  It is not allowed by our group security rule.

**Comment by hannwang1** [comment]
> And meanwhile, we found a root cause and fix/workaround: seems that command tries to talk to legacy AAD Graph API and resolve objectIDs, and this return Forbidden error
> We added permission `Directory.Read.All` application permission to app registration and it started working.
> It's kind of shame that Az cmdlets still use the legacy API that has been discouraged for few years now. Also, the documentation could be more specific in terms of level of permission needed to be able to successfully run the cmdlet.

@jformacek, the  `Directory.Read.All` permission you added is Legacy 'Azure Active Directory Graph', right?  In my org, it is disabled.  I cannot even select it.  So I am totally blocked with no workaround.

![image](https://user-images.githubusercontent.com/44303608/134764039-4095506b-c7d0-404a-96a8-09421d3f15ac.png)

**Comment by hannwang1** [comment]
I also tried @jikuja workaround, provided all parameters, not working for me either.

**Comment by jikuja** [comment]
I did a new workaround: a function that calls `Invoke-AzRestMethod` directly for use cases when are required information are known by the caller. The first workaround I offered did not work because Cmdlet has bug(s) that causes following: AD graph API is queried even all the required information is given by the user.

https://docs.microsoft.com/en-us/azure/role-based-access-control/role-assignments-rest has a good explanation how to do role assignments with REST API if you need a working workaround ASAP without adding permissions to your service principal.

-------

> Are you thinking we should use graph.microsoft.com?

I hope some work is already done because:

> Azure Active Directory (Azure AD) Graph is deprecated. To avoid loss of functionality, migrate your applications to Microsoft Graph before June 30, 2022 when Azure AD Graph API endpoints will stop responding to requests.

(ref: https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-overview)

I did not find any bug report reports or PRs about the Microsoft Graph migration. Did I miss those?

**Comment by hannwang1** [comment]
@jikuja I will try call API.  the cmdlet New-AzRoleAssignment, Get-AzRoleAssignment do not work for me.
I cannot add Microsoft. Graph Directory.read.all permission for my DevOps SPN.  It is not allowed by my org security rule.

**Comment by jikuja** [comment]
This thread has now multiple problems:

1. Using the API that is being deprecated soon
  * > Microsoft will continue technical support and apply security fixes for Azure AD Graph until June 30, 2022 when all functionality and support will end. 
  * Is there already a dedicated ticket, plans, or ETA for a rewrite? 
2. Using the API that requires extra permission
3. Having a logic bug that makes requests to deprecated API even when not needed
  * I hope this is fixed when migrating to Microsoft Graph
4. Having old workaround not working anymore

**Comment by dingmeng-xue** [comment]
We are asking MS Identity team to provide the workaround. Azure PowerShell module doesn't support to add or grant API permission. I find AzureAd module support adding permission but no grant. For the time being, the only way in my mind is leveraging Azure CLI [az ad app permission add](https://docs.microsoft.com/en-us/cli/azure/ad/app/permission?view=azure-cli-latest#az_ad_app_permission_add) to add permission of AAD Graph. The value of API id and permission id are obscure. You can reference https://github.com/mjisaak/azure-active-directory/blob/master/README.md#well-known-appids  

I will check whether MSGraph module can support it. Will keep you updated.

@jikuja , @cutecycle , Thanks for your input. I have created a request to track New-AzRoleAssignment fix https://github.com/Azure/azure-powershell/issues/16054

**Comment by KranthiPakala-MSFT** [comment]
@wmengmsft , @MehaKaushik , @shurd, @anfeldma-ms - Could you please have a look at this issue and share your thoughts. Thank you

**Comment by anaismiller** [comment]
There is also this issue with Remove-AzRoleAssignment. Is there a known workaround without querying Graph API like the New-AzRoleAssignment query provided by jikuja?

**Comment by catalinpopa247** [comment]
@dingmeng-xue did anything change recently? Seems like the workaround you mentioned is not valid anymore. It fails with the following error: 

Set-AzureADApplication : Error occurred while executing SetApplication
Code: Request_BadRequest
Message: Property requiredResourceAccess.resourceAccess is invalid.
RequestId: 9617d8c8-febc-44cc-a57b-9d9c197558bc
DateTimeStamp: Fri, 19 Nov 2021 09:47:16 GMT
Details: PropertyName  - requiredResourceAccess.resourceAccess, PropertyErrorCode  - GenericError
HttpStatusCode: BadRequest
HttpStatusDescription: Bad Request
HttpResponseStatus: Completed
At line:1 char:1
+ Set-AzureADApplication -ObjectId $AppObjectID -RequiredResourceAccess ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Set-AzureADApplication], ApiException
    + FullyQualifiedErrorId : Microsoft.Open.AzureAD16.Client.ApiException,Microsoft.Open.AzureAD16.PowerShell.SetAppl
   ication
   
  
Did anything change? Is the migration from Azure Graph API to Microsoft Graph complete?

**Comment by dingmeng-xue** [comment]
Hi @catalinpopa247 , is it possible that you share the debug log of AzureAD cmdlet? The module is not owned by Azure PowerShell team. We didn't know whether there is anything change.

**Comment by jasonfritts** [comment]
Try updating Az.Resources module to the latest version which uses graph.microsoft.com instead of graph.windows.net

https://www.powershellgallery.com/packages/Az.Resources/5.1.0

**Comment by LIYINDUO** [comment]
I also encountered this issue “Exception of type Microsoft.Rest.Azure.CloudException was thrown”.
Azure Active Directory Graph API permission assigment feature via Azure Portal seems to be closed.
So a more manually approach is applicable for assigning this permission. Below script worked for me.
Kindly find the "object id" of the service pricinple you are using to call "New-AzRoleAssignment" in enterprise application panel(Not in app registrations). Then use this value for the $miObjectID parameter: 

Connect-AzureAD
$miObjectID = "" # Object Id
$appId = "00000002-0000-0000-c000-000000000000"  # Graph Id
$permissionsToAdd = "Directory.Read.All"
$app = Get-AzureADServicePrincipal -Filter "AppId eq '$appId'"
$role = $app.AppRoles | where Value -Like $permissionsToAdd | Select-Object -First 1
New-AzureADServiceAppRoleAssignment -Id $role.Id -ObjectId $miObjectID -PrincipalId $miObjectID -ResourceId $app.ObjectId


---
Not sure if the above workaround is recommeded though. Switch to assign role via  Azure CLI could be a better way to go around the issue for now.

**Comment by dingmeng-xue** [comment]
You also can use the latest version of Az.Resources 5.x. All related features are migrated to MSGraph API.

**Comment by jformacek** [comment]
Tested with new module and works. Closing the issue.

**Comment by nExoRek** [comment]
it does not return an error but it does not return sigininname nor displayname either /:

**Comment by dingmeng-xue** [comment]
@nExoRek , could you create a new Github issue and post script and debug message? You can share debug message with setting `$DebugPreference = "Continue"`
